### PR TITLE
[StatsMigration] Create PartitionClassStatsPublisher in StatsManager

### PR DESCRIPTION
**Pull request by justinlin-linkedin**
Adding PartitionClassStatsPublisher in StatsManager.
Adding integration test for StatsManager to publish stats to mysql database.

### CODE REVIEW

File: ambry-server/src/main/java/com/github/ambry/server/StatsManager.java [hunk 1]
```diff
@@ -474,16 +521,67 @@ public void run() {
...
+      }
+    }
+
+    void cancel() {
+      cancelled = true;
+    }
+  }
+
+  /**
+   * Runnable calss that collects, publishes partition class stats to mysql database.
```
Comment (jsjtzyy) [inline ambry-server/src/main/java/com/github/ambry/server/StatsManager.java:544]: typo: `class`

**Review by jsjtzyy — APPROVED** [review]
lgtm after minor comments are addressed.

File: ambry-api/src/main/java/com/github/ambry/config/StatsManagerConfig.java [hunk 2]
```diff
@@ -74,6 +77,21 @@
...
+   * this configuration.
+   */
+  @Config(STATS_ENABLE_PARTITION_CLASS_MYSQL_REPORT)
+  @Default("false")
+  public final boolean enablePartitionClassMysqlReport;
+
+  /**
+   * The time period in seconds that configures how often partition class stats are published to mysql.
+   */
+  @Config(STATS_PUBLISH_PARTITION_CLASS_REPORT_PERIOD_IN_SECS)
```
[the remark this answers is not in the document]
Comment (justinlin-linkedin) [reply ambry-api/src/main/java/com/github/ambry/config/StatsManagerConfig.java:91]: sure.

File: ambry-server/src/main/java/com/github/ambry/server/StatsManager.java [hunk 3]
```diff
@@ -129,18 +131,31 @@
...
-    scheduler.scheduleAtFixedRate(statsAggregator, actualDelay, publishPeriodInSecs, TimeUnit.SECONDS);
+    accountsStatsPublisher = new AccountStatsPublisher(accountStatsMySqlStore);
+    int actualDelay = config.initialDelayUpperBoundInSecs > 0 ? ThreadLocalRandom.current()
+        .nextInt(config.initialDelayUpperBoundInSecs) : 0;
+    logger.info("Scheduling account stats publishing job with an initial delay of {} secs", actualDelay);
+    scheduler.scheduleAtFixedRate(accountsStatsPublisher, actualDelay, config.publishPeriodInSecs, TimeUnit.SECONDS);
+
+    if (config.enablePartitionClassMysqlReport) {
+      partitionClassStatsPublisher = new PartitionClassStatsPublisher(accountStatsMySqlStore);
+      long initalDelay = ThreadLocalRandom.current().nextLong(config.publishPartitionClassReportPeriodInSecs / 2)
```
[the remark this answers is not in the document]
Comment (justinlin-linkedin) [reply ambry-server/src/main/java/com/github/ambry/server/StatsManager.java]: the logic is that we always wait for at least half of the interval. If the interval is 1 hour, then at least we will wait for 30 minutes.

File: ambry-server/src/main/java/com/github/ambry/server/StatsManager.java [hunk 4]
```diff
@@ -209,6 +224,37 @@ void collectAndAggregate(StatsSnapshot aggregatedSnapshot, PartitionId partition
...
+    Store store = storageManager.getStore(partitionId, false);
+    if (store == null) {
+      unreachablePartitions.add(partitionId);
+    } else {
+      try {
+        long fetchAndAggregatePerStoreStartTimeMs = time.milliseconds();
+        StoreStats storeStats = store.getStoreStats();
+        Map<StatsReportType, StatsSnapshot> snapshotsByType =
+            storeStats.getStatsSnapshots(EnumSet.of(StatsReportType.PARTITION_CLASS_REPORT), time.milliseconds(),
+                publishExcludeAccountIds);
```
[the remark this answers is not in the document]
Comment (justinlin-linkedin) [reply ambry-server/src/main/java/com/github/ambry/server/StatsManager.java:244]: we can do that with current dataset. In aggregated account reports, we have account level data as well.

File: ambry-server/src/main/java/com/github/ambry/server/StatsManager.java [hunk 5]
```diff
@@ -435,53 +478,108 @@ long getPermanentDeleteTombstoneSize() {
...
-          StatsWrapper statsWrapper = new StatsWrapper(statsHeader, aggregatedSnapshot);
+      Object lock = accountStatsMySqlStore != null ? accountStatsMySqlStore : this;
+      synchronized (lock) {
+        logger.info("Aggregating account stats for local report");
+        try {
+          resetDeleteTombstoneStats();
+          long totalFetchAndAggregateStartTimeMs = time.milliseconds();
+          StatsSnapshot aggregatedSnapshot = new StatsSnapshot(0L, new HashMap<>());
+          List<PartitionId> unreachablePartitions = new ArrayList<>();
+          Iterator<PartitionId> iterator = (new HashSet<>(partitionToReplicaMap.keySet())).iterator();
```
Comment (lightningrob) [inline ambry-server/src/main/java/com/github/ambry/server/StatsManager.java:501]: If partitionToReplicaMap is never modified after the constructor, why do you need to copy the key set before iterating?

File: ambry-server/src/main/java/com/github/ambry/server/StatsManager.java [hunk 6]
```diff
@@ -435,53 +478,108 @@ long getPermanentDeleteTombstoneSize() {
...
-        updateAggregatedDeleteTombstoneStats();
-        metrics.initDeleteStatsGaugesIfNeeded();
-        if (!cancelled) {
-          metrics.totalFetchAndAggregateTimeMs.update(time.milliseconds() - totalFetchAndAggregateStartTimeMs);
-          StatsHeader statsHeader = new StatsHeader(StatsHeader.StatsDescription.STORED_DATA_SIZE, time.milliseconds(),
-              partitionToReplicaMap.keySet().size(), partitionToReplicaMap.keySet().size() - unreachableStores.size(),
-              unreachableStores);
-          // First write to account stats
-          StatsWrapper statsWrapper = new StatsWrapper(statsHeader, aggregatedSnapshot);
+      Object lock = accountStatsMySqlStore != null ? accountStatsMySqlStore : this;
```
Comment (lightningrob) [inline ambry-server/src/main/java/com/github/ambry/server/StatsManager.java:493]: Probably simpler to use synchronized(this) in both cases.

File: ambry-server/src/main/java/com/github/ambry/server/StatsManager.java [hunk 7]
```diff
@@ -435,53 +478,108 @@ long getPermanentDeleteTombstoneSize() {
...
+      synchronized (lock) {
+        logger.info("Aggregating account stats for local report");
+        try {
+          resetDeleteTombstoneStats();
```
Comment (lightningrob) [inline ambry-server/src/main/java/com/github/ambry/server/StatsManager.java:497]: Could you add a few more inline comments explaining what each of the steps does?  It's a lot of code with relatively few comments.

File: ambry-server/src/main/java/com/github/ambry/server/StatsManager.java [hunk 8]
```diff
@@ -435,53 +478,108 @@ long getPermanentDeleteTombstoneSize() {
...
+    private volatile boolean cancelled = false;
+    private final AccountStatsMySqlStore accountStatsMySqlStore;
+
+    PartitionClassStatsPublisher(AccountStatsMySqlStore accountStatsMySqlStore) {
+      this.accountStatsMySqlStore = accountStatsMySqlStore;
+    }
+
+    @Override
+    public void run() {
+      synchronized (accountStatsMySqlStore) {
```
Comment (lightningrob) [inline ambry-server/src/main/java/com/github/ambry/server/StatsManager.java:555]: Why can we assume that accountStatsMySqlStore is non-null here but not in the other task?
Also, do the actions of PartitionClassStatsPublisher and AccountStatsPublisher need to be serialized with each other (using same lock)?

File: ambry-server/src/main/java/com/github/ambry/server/StatsManager.java [hunk 9]
```diff
@@ -435,53 +478,108 @@ long getPermanentDeleteTombstoneSize() {
...
+        logger.info("Aggregating partition class stats for local report");
+        try {
+          long totalFetchAndAggregateStartTimeMs = time.milliseconds();
+          StatsSnapshot aggregatedSnapshot = new StatsSnapshot(0L, new HashMap<>());
+          List<PartitionId> unreachablePartitions = new ArrayList<>();
+          Iterator<PartitionId> iterator = (new HashSet<>(partitionToReplicaMap.keySet())).iterator();
```
Comment (lightningrob) [inline ambry-server/src/main/java/com/github/ambry/server/StatsManager.java:561]: Same question about partitionToReplicaMap.

File: ambry-server/src/main/java/com/github/ambry/server/StatsManager.java [hunk 10]
```diff
@@ -435,53 +478,108 @@ long getPermanentDeleteTombstoneSize() {
...
+            metrics.totalFetchAndAggregateTimeMs.update(time.milliseconds() - totalFetchAndAggregateStartTimeMs);
+            StatsHeader statsHeader =
+                new StatsHeader(StatsHeader.StatsDescription.STORED_DATA_SIZE, time.milliseconds(),
+                    partitionToReplicaMap.keySet().size(),
+                    partitionToReplicaMap.keySet().size() - unreachableStores.size(), unreachableStores);
+            StatsWrapper statsWrapper = new StatsWrapper(statsHeader, aggregatedSnapshot);
+            accountStatsMySqlStore.storePartitionClassStats(statsWrapper);
+            logger.info("Local partition class stats snapshot published to {}", statsOutputFile.getAbsolutePath());
+          }
+        } catch (Exception | Error e) {
```
Comment (lightningrob) [inline ambry-server/src/main/java/com/github/ambry/server/StatsManager.java:579]: Throwable is equivalent.  But if we get an Error like OOM, do we really want to eat it?

File: ambry-server/src/integration-test/java/com/github/ambry/server/StatsManagerIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+
+  /**
+   * Test partition class stats publisher
+   * @throws Exception
+   */
+  @Test
+  public void testPartitionClassStatsPublisher() throws Exception {
+    // Before publishing partition class stats, we have to publish account stats since they use the same data
+    StatsManager.AccountStatsPublisher accountStatsPublisher =
+        statsManager.new AccountStatsPublisher(accountStatsMySqlStore);
```
Comment (lightningrob) [inline ambry-server/src/integration-test/java/com/github/ambry/server/StatsManagerIntegrationTest.java:137]: Should also test this will null store since that is an option.

File: ambry-server/src/main/java/com/github/ambry/server/StatsManager.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (lightningrob) [reply ambry-server/src/main/java/com/github/ambry/server/StatsManager.java:241]: Please avoid mention of LI things in github reviews.

**Review by lightningrob — COMMENTED** [review]
Logic looks good.  A few comments mainly about concurrency.

File: ambry-server/src/main/java/com/github/ambry/server/StatsManager.java (hunk 5 above)
Comment (justinlin-linkedin) [reply ambry-server/src/main/java/com/github/ambry/server/StatsManager.java:501]: updated.

File: ambry-server/src/main/java/com/github/ambry/server/StatsManager.java (hunk 6 above)
Comment (justinlin-linkedin) [reply ambry-server/src/main/java/com/github/ambry/server/StatsManager.java:493]: I want to synchronize on accountStatsMysqlStore since accountStatsMysqlStore only has one connection and multiple threads accessing this connection is not safe. That's why every time I want to use accountStatsMysqlStore, I would have to lock this object.
There are other threads trying to use accountStatsMysqlStore, one of them is MysqlReportAggregatorTask. This is a aggregation task that read all the stats reports and aggregate them and then save the aggregated result back to mysql database. By synchronizing on accountStatsMysqlStore,  we can be sure that there is no concurrency issue with accessing the single connection in accountStatsMysqlStore.
However, the accountStatsMysqlStore might be null, it's not yet enabled everywhere.

File: ambry-server/src/main/java/com/github/ambry/server/StatsManager.java (hunk 7 above)
Comment (justinlin-linkedin) [reply ambry-server/src/main/java/com/github/ambry/server/StatsManager.java:497]: updated.

File: ambry-server/src/main/java/com/github/ambry/server/StatsManager.java (hunk 8 above)
Comment (justinlin-linkedin) [reply ambry-server/src/main/java/com/github/ambry/server/StatsManager.java:555]: This is guaranteed, since we have to enable mysql report before we can enable the partition class reports. We have a check in StatsManagerConfig.java.

File: ambry-server/src/main/java/com/github/ambry/server/StatsManager.java (hunk 9 above)
Comment (justinlin-linkedin) [reply ambry-server/src/main/java/com/github/ambry/server/StatsManager.java:561]: updated.

File: ambry-server/src/main/java/com/github/ambry/server/StatsManager.java (hunk 10 above)
Comment (justinlin-linkedin) [reply ambry-server/src/main/java/com/github/ambry/server/StatsManager.java:579]: Will ignore Error.

File: ambry-server/src/integration-test/java/com/github/ambry/server/StatsManagerIntegrationTest.java (hunk 11 above)
Comment (justinlin-linkedin) [reply ambry-server/src/integration-test/java/com/github/ambry/server/StatsManagerIntegrationTest.java:137]: This integration test is mainly for testing mysql database. In unit test, we consider the mysql database doesn't exist by default. So we have to do the test for mysql database here. The case when accountStatsMysqlStore is null, has already been tested in unit tests.

File: ambry-server/src/main/java/com/github/ambry/server/StatsManager.java (hunk 6 above)
Comment (lightningrob) [reply ambry-server/src/main/java/com/github/ambry/server/StatsManager.java:493]: Okay, makes sense.

**Review by lightningrob — APPROVED** [review]
LGTM
